DataDog/dd-trace-js
# [serverless] Add DynamoDB Span Pointers

*Large PR, but 1000+ lines are tests. Only ~250 lines are actual code from this PR. All these tests are necessary to test all the different possible cases. Please review commit by commit to make the review process easier.*

### What does this PR do?
Adds span pointers in DynamoDB for `putItem`, `updateItem`, `deleteItem`, `transactWriteItems`, and `batchWriteItem` requests.

For more info on the details of how span pointers work for each of these operations, see https://github.com/DataDog/dd-span-pointer-rules/blob/main/AWS/DynamoDB/Item/README.md.

Span pointers are similar to [Span Links](https://docs.datadoghq.com/tracing/trace_collection/span_links/), but for cases when it is impossible to pass the Trace ID and Span ID between the spans that need to be linked.

When the calculated hashes for the upstream and downstream lambdas match, the Datadog backend will automatically link the two traces together.

<img width="1038" alt="Screenshot 2024-11-19 at 4 41 31 PM" src="https://github.com/user-attachments/assets/1cc5e0d7-741d-47db-96a8-6aa074a655a7">
When clicking on the linked span, a new tab opens linking to the downstream Lambda function that was triggered by this Dynamo object update.

### Motivation
This feature already exists in Python, and I'm working on adding it to all other runtimes (Node, .NET, Java, Golang).

### Additional Notes
- This handles the 'upstream' case, for Lambdas that make Dynmo requests. I will have another PR in datadog-lambda-js to handle the 'downstream' case, for Lambdas that are triggered by Dynamo updates.
- My changes in `packages/dd-trace/src/util.js` are located there because it is a public API, since `datadog-lambda-js` will reuse the exported fields. 
- `UpdateItem` and `DeleteItem` require the primary key(s) as parameters, which makes our lives much easier. However, for PutItem, we have no way of determining which keys are the primary keys, so the user must set an environment variable specifying the primary key(s).
- I add the telemetry key to Datadog backend in https://github.com/DataDog/dd-go/pull/161069

### Testing Guidelines

Easy: Checkout [this span](https://ddserverless.datadoghq.com/functions?cloud=aws&config_serverless-auto-span-links=true&config_trace_show_hidden_metadata=true&entity_view=lambda_functions&fromUser=false&graphType=flamegraph&my_code=enabled&panel_end=1732052495436&panel_paused=false&panel_start=1732038095436&panel_tab=invocations&shouldShowLegend=true&sort=time&sp=%5B%7B%22p%22%3A%7B%22entityId%22%3A%22aws-lambda-functions%2Bnhulston-javascript-upstream-dynamo-dev-hello%2Bus-east-1%2B425362996713%22%7D%2C%22i%22%3A%22lambda-panel%22%7D%2C%7B%22p%22%3A%7B%22traceID%22%3A%222053206778893931889%22%2C%22selectedSpanID%22%3A%222053206778893931889%22%7D%2C%22i%22%3A%22trace-panel%22%7D%5D&spanID=2053206778893931889&spanViewType=span-links&text_search=nhulston-javascript&traceID=2053206778893931889&start=1732048352571&end=1732051952571&paused=false), enable the feature flag, and you will see that it's pointing to the downstream Lambda.

Thorough testing: Run [this Lambda function](https://us-east-1.console.aws.amazon.com/lambda/home?region=us-east-1#/functions/nhulston-javascript-upstream-dynamo-dev-hello?subtab=envVars&tab=testing) with the event payload
```json
{
  "shouldUseV2": false,
  "twoKeyTable": false,
  "shouldPutItem": false,
  "shouldUpdateItem": false,
  "shouldDeleteItem": false,
  "shouldBatchWriteItems": false,
  "shouldTransactWriteItems": false
}
```
and change one of the bools to true. Also, try both with AWS SDK v2 and v3, and try with twoKeyTable vs oneKeyTable. Enable the span pointers feature flag, and check Datadog to ensure that the spans are properly linked.

I also added tests:
```shell
# Unit tests
yarn test packages/dd-trace/test/util.spec.js

# Integration tests
docker run -d -p 4566:4566 localstack/localstack
PLUGINS=aws-sdk SERVICES=localstack DD_DATA_STREAMS_ENABLED=true yarn test:plugins:ci 
```
